File: lineageweave/project_history.py
"""Build evidence-bound project histories from already-authorized rows.

Callers must apply RBAC, ABAC, source eligibility, and knowledge-cutoff
filtering before invoking this module. The pure projection layer then orders
visible source records, preserves explicit and semantic project evidence,
compares observed responsibility evidence, and exposes persisted lineage as
related history without promoting it to causality or an HR assignment ledger.
"""

from __future__ import annotations

from collections import deque
from collections.abc import Mapping, Sequence
from datetime import datetime, timezone
from decimal import Decimal
import math
from typing import Any
from unicodedata import normalize

PROJECT_HISTORY_CONTRACT_VERSION = 1
PROJECT_HISTORY_TIME_BASIS = "source_post_created_at_fallback"
PROJECT_HISTORY_DOCUMENT_TIME_BASIS = "document_time"
PROJECT_HISTORY_MAX_DEPTH = 8
PROJECT_HISTORY_MAX_PATHS_PER_EVENT = 32

_VOC_CODES = frozenset({"voc", "vocc", "voco", "vom", "vop"})
_TRUTH_ORDER = {"observed": 0, "inferred": 1}
_DISPLAY_NAME_ORDER = {"source_project_name": 0, "semantic_project_name": 1}


def normalize_project_key(value: str) -> str:
    """Return the exact project-identity comparison key.

    Unicode compatibility normalization lets full-width and compatibility
    forms match without introducing fuzzy identity. Empty and oversized keys
    fail closed.
    """

    normalized = normalize("NFKC", value).strip().lower()
    if not normalized:
        raise ValueError("project key must not be empty")
    if len(normalized.encode("utf-8")) > 256:
        raise ValueError("project key exceeds 256 UTF-8 bytes")
    return normalized


def classify_project_event(
    *,
    title: str,
    source_stage_code: str | None,
    source_detail_state_code: str | None,
    voc_type_code: str | None,
    is_focus: bool,
) -> str:
    """Return a non-authoritative display classification for one source row.

    Only the persisted controlled VOC code is classified. Free-text titles,
    stages, and detail states remain evidence fields; this projection never
    guesses lifecycle semantics from words. ``is_focus`` is retained for
    contract compatibility but never changes the truth status or creates an
    event.
    """

    del is_focus
    del title, source_stage_code, source_detail_state_code
    if (voc_type_code or "").strip().lower() in _VOC_CODES:
        return "voc_received"
    return "source_recorded"


def responsibility_transition_code(
    previous_actor_keys: Sequence[str], current_actor_keys: Sequence[str]
) -> str:
    """Compare adjacent observed responsibility evidence.

    Missing evidence on either row is an ``assignment_gap`` evidence state,
    not proof of an operational or HR vacancy. Equal non-empty actor sets are
    continuous; different non-empty sets are a handoff.
    """

    previous = frozenset(key for key in previous_actor_keys if key)
    current = frozenset(key for key in current_actor_keys if key)
    if not previous or not current:
        return "assignment_gap"
    if previous == current:
        return "continuous"
    return "handoff"


def _as_utc(value: datetime) -> str:
    """Serialize a source clock as canonical UTC RFC 3339 text."""

    aware = value if value.tzinfo is not None else value.replace(tzinfo=timezone.utc)
    return aware.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def _actor_key(role: Mapping[str, Any]) -> str:
    """Return a stable key for one observed role actor."""

    catalog_fields = (
        ("person", role.get("cataloged_person_id")),
        ("team", role.get("cataloged_team_id")),
        ("organization", role.get("cataloged_corporate_entity_id")),
    )
    for prefix, value in catalog_fields:
        if value:
            return f"{prefix}:{value}"
    parts = (
        str(role.get("actor_type_code") or "unknown"),
        str(role.get("actor_name") or ""),
        str(role.get("affiliated_organization_name") or ""),
    )
    return "text:" + "\u001f".join(normalize("NFKC", part).strip().lower() for part in parts)


def _score(value: object) -> float:
    """Return a finite JSON-compatible lineage score."""

    if isinstance(value, bool) or not isinstance(value, (int, float, Decimal)):
        raise ValueError("lineage score must be numeric")
    result = float(value)
    if math.isnan(result) or result in (float("inf"), float("-inf")):
        raise ValueError("lineage score must be finite")
    return result


def _prior_paths(
    ordered_event_ids: Sequence[str],
    edge_rows: Sequence[Mapping[str, Any]],
    *,
    maximum_depth: int,
    maximum_paths_per_event: int,
) -> dict[str, list[dict[str, Any]]]:
    """Return deterministic shortest visible predecessor paths per event."""

    event_index = {event_id: index for index, event_id in enumerate(ordered_event_ids)}
    reverse_edges: dict[str, list[dict[str, Any]]] = {event_id: [] for event_id in ordered_event_ids}
    for row in edge_rows:
        parent = str(row["parent_post_id"])
        child = str(row["child_post_id"])
        if parent not in event_index or child not in event_index:
            continue
        if event_index[parent] >= event_index[child]:
            continue
        reverse_edges[child].append(
            {
                "parent_event_id": parent,
                "child_event_id": child,
                "fused_score": _score(row["fused_score"]),
            }
        )
    for edges in reverse_edges.values():
        edges.sort(key=lambda edge: (event_index[edge["parent_event_id"]], edge["parent_event_id"]))

    result: dict[str, list[dict[str, Any]]] = {}
    for target in ordered_event_ids:
        queue: deque[tuple[str, tuple[str, ...], tuple[dict[str, Any], ...]]] = deque(
            [(target, (target,), ())]
        )
        best_depth = {target: 0}
        paths: list[dict[str, Any]] = []
        while queue and len(paths) < maximum_paths_per_event:
            current, reverse_event_path, reverse_edge_path = queue.popleft()
            depth = len(reverse_edge_path)
            if depth >= maximum_depth:
                continue
            for edge in reverse_edges[current]:
                parent = edge["parent_event_id"]
                if parent in reverse_event_path:
                    continue
                next_depth = depth + 1
                if best_depth.get(parent, maximum_depth + 1) <= next_depth:
                    continue
                best_depth[parent] = next_depth
                next_events = reverse_event_path + (parent,)
                next_edges = reverse_edge_path + (edge,)
                ordered_events = list(reversed(next_events))
                ordered_edges = list(reversed(next_edges))
                paths.append(
                    {
                        "source_event_id": parent,
                        "target_event_id": target,
                        "event_ids": ordered_events,
                        "edges": ordered_edges,
                        "minimum_fused_score": min(item["fused_score"] for item in ordered_edges),
                        "truth_status_code": "inferred",
                        "source_relation_code": "post_lineage_edge",
                        "provenance": "post_lineage_edge.fused_score",
                    }
                )
                queue.append((parent, next_events, next_edges))
                if len(paths) >= maximum_paths_per_event:
                    break
        paths.sort(
            key=lambda path: (
                len(path["edges"]),
                event_index[path["source_event_id"]],
                tuple(path["event_ids"]),
            )
        )
        result[target] = paths
    return result


def build_project_history_projection(
    *,
    project_key: str,
    focus_event_id: str | None,
    event_rows: Sequence[Mapping[str, Any]],
    match_rows: Sequence[Mapping[str, Any]],
    role_rows: Sequence[Mapping[str, Any]],
    edge_rows: Sequence[Mapping[str, Any]],
    truncated: bool = False,
    maximum_depth: int = PROJECT_HISTORY_MAX_DEPTH,
    maximum_paths_per_event: int = PROJECT_HISTORY_MAX_PATHS_PER_EVENT,
) -> dict[str, Any]:
    """Build the versioned project-history projection.

    Inputs must already be visible, eligible, and within the requested cutoff.
    Duplicate source rows and role rows are collapsed deterministically. An
    observed source project name outranks an inferred semantic display name.
    """

    normalized_key = normalize_project_key(project_key)
    if maximum_depth < 1 or maximum_depth > PROJECT_HISTORY_MAX_DEPTH:
        raise ValueError("maximum_depth is outside the supported bound")
    if maximum_paths_per_event < 1 or maximum_paths_per_event > PROJECT_HISTORY_MAX_PATHS_PER_EVENT:
        raise ValueError("maximum_paths_per_event is outside the supported bound")

    deduplicated: dict[str, Mapping[str, Any]] = {}
    for row in event_rows:
        event_id = str(row["post_id"])
        current = deduplicated.get(event_id)
        row_clock = row.get("event_occurred_at") or row["created_at"]
        current_clock = (
            current.get("event_occurred_at") or current["created_at"]
            if current is not None
            else None
        )
        if current is None or (row_clock, row["created_at"], event_id) < (
            current_clock,
            current["created_at"],
            event_id,
        ):
            deduplicated[event_id] = row
    ordered_rows = sorted(
        deduplicated.values(),
        key=lambda row: (
            row.get("event_occurred_at") or row["created_at"],
            row["created_at"],
            str(row["post_id"]),
        ),
    )
    if not ordered_rows:
        raise ValueError("project history requires at least one visible event")
    ordered_ids = [str(row["post_id"]) for row in ordered_rows]
    event_index = {event_id: index for index, event_id in enumerate(ordered_ids)}
    effective_focus = focus_event_id or ordered_ids[-1]
    if effective_focus not in event_index:
        raise ValueError("focus event is not in the visible project history")

    matches_by_event: dict[str, list[dict[str, Any]]] = {event_id: [] for event_id in ordered_ids}
    display_names: list[tuple[int, int, str, str]] = []
    seen_matches: set[tuple[str, str, str]] = set()
    for row in match_rows:
        event_id = str(row["post_id"])
        if event_id not in matches_by_event:
            continue
        matched_value = str(row["matched_value"])
        if normalize_project_key(matched_value) != normalized_key:
            continue
        kind = str(row["match_kind_code"])
        key = (event_id, kind, matched_value)
        if key in seen_matches:
            continue
        seen_matches.add(key)
        confidence = row.get("confidence")
        if confidence is not None:
            confidence = _score(confidence)
        truth = "observed" if kind.startswith("source_") else "inferred"
        matches_by_event[event_id].append(
            {
                "match_kind_code": kind,
                "matched_value": matched_value,
                "truth_status_code": truth,
                "confidence": confidence,
                "ontology_iri": row.get("ontology_iri"),
                "provenance": str(row["provenance"]),
            }
        )
        if kind in _DISPLAY_NAME_ORDER:
            display_names.append(
                (
                    _DISPLAY_NAME_ORDER[kind],
                    event_index[event_id],
                    normalize_project_key(matched_value),
                    matched_value,
                )
            )
    for matches in matches_by_event.values():
        matches.sort(
            key=lambda item: (
                _TRUTH_ORDER[item["truth_status_code"]],
                item["match_kind_code"],
                item["matched_value"],
            )
        )

    roles_by_event: dict[str, list[dict[str, Any]]] = {event_id: [] for event_id in ordered_ids}
    actor_keys_by_event: dict[str, list[str]] = {event_id: [] for event_id in ordered_ids}
    distinct_actor_keys: set[str] = set()
    seen_roles: set[tuple[str, str, str]] = set()
    for row in role_rows:
        event_id = str(row["post_id"])
        if event_id not in roles_by_event:
            continue
        actor_key = _actor_key(row)
        responsibility = str(row["responsibility"])
        role_key = (event_id, actor_key, responsibility)
        if role_key in seen_roles:
            continue
        seen_roles.add(role_key)
        distinct_actor_keys.add(actor_key)
        actor_keys_by_event[event_id].append(actor_key)
        roles_by_event[event_id].append(
            {
                "actor_key": actor_key,
                "actor_name": str(row["actor_name"]),
                "actor_type_code": str(row["actor_type_code"]),
                "affiliated_organization_name": row.get("affiliated_organization_name"),
                "responsibility": responsibility,
                "truth_status_code": "observed",
                "provenance": "post_summary_role",
            }
        )
    for event_id, roles in roles_by_event.items():
        roles.sort(key=lambda role: (role["actor_type_code"], role["actor_name"], role["actor_key"]))
        actor_keys_by_event[event_id] = sorted(set(actor_keys_by_event[event_id]))

    paths_by_event = _prior_paths(
        ordered_ids,
        edge_rows,
        maximum_depth=maximum_depth,
        maximum_paths_per_event=maximum_paths_per_event,
    )

    events: list[dict[str, Any]] = []
    previous_actor_keys: Sequence[str] | None = None
    for row in ordered_rows:
        event_id = str(row["post_id"])
        event_occurred_at = row.get("event_occurred_at")
        time_basis_code = (
            PROJECT_HISTORY_DOCUMENT_TIME_BASIS
            if event_occurred_at is not None
            else PROJECT_HISTORY_TIME_BASIS
        )
        current_actor_keys = actor_keys_by_event[event_id]
        transition = (
            None
            if previous_actor_keys is None
            else responsibility_transition_code(previous_actor_keys, current_actor_keys)
        )
        events.append(
            {
                "event_id": event_id,
                "source_post_id": event_id,
                "event_title": str(row["post_title"]),
                "event_type_code": classify_project_event(
                    title=str(row["post_title"]),
                    source_stage_code=row.get("source_stage_code"),
                    source_detail_state_code=row.get("source_detail_state_code"),
                    voc_type_code=row.get("voc_type_code"),
                    is_focus=event_id == effective_focus,
                ),
                "event_type_basis_code": "controlled_source_code",
                "occurred_at": _as_utc(event_occurred_at or row["created_at"]),
                "time_basis_code": time_basis_code,
                "voc_type_code": row.get("voc_type_code"),
                "source_stage_code": row.get("source_stage_code"),
                "source_detail_state_code": row.get("source_detail_state_code"),
                "project_matches": matches_by_event[event_id],
                "observed_responsibilities": roles_by_event[event_id],
                "responsibility_transition_code": transition,
                "related_prior_paths": paths_by_event[event_id],
            }
        )
        previous_actor_keys = current_actor_keys

    project_name = min(display_names)[3] if display_names else project_key.strip()
    return {
        "contract_version": PROJECT_HISTORY_CONTRACT_VERSION,
        "project_key": project_key.strip(),
        "normalized_project_key": normalized_key,
        "project_name": project_name,
        "focus_event_id": effective_focus,
        "time_basis_code": (
            PROJECT_HISTORY_DOCUMENT_TIME_BASIS
            if all(event["time_basis_code"] == PROJECT_HISTORY_DOCUMENT_TIME_BASIS for event in events)
            else PROJECT_HISTORY_TIME_BASIS
        ),
        "event_count": len(events),
        "distinct_observed_actor_count": len(distinct_actor_keys),
        "truncated": bool(truncated),
        "events": events,
    }

File: frontend/src/projectHistory.ts
import type { ProjectEvidence } from "./api";
import type { Locale } from "./i18n";

export type ProjectHistoryTruthStatus = "observed" | "inferred";
export type ResponsibilityTransitionCode = "continuous" | "handoff" | "assignment_gap";
export type ProjectHistoryTimeBasis = "source_post_created_at_fallback" | "document_time";

export interface ProjectHistoryMatch {
  match_kind_code: string;
  matched_value: string;
  truth_status_code: ProjectHistoryTruthStatus;
  confidence: number | null;
  ontology_iri: string | null;
  provenance: string;
}

export interface ProjectHistoryResponsibility {
  actor_key: string;
  actor_name: string;
  actor_type_code: string;
  affiliated_organization_name: string | null;
  responsibility: string;
  truth_status_code: "observed";
  provenance: string;
}

export interface ProjectHistoryPathEdge {
  parent_event_id: string;
  child_event_id: string;
  fused_score: number;
}

export interface ProjectHistoryPriorPath {
  source_event_id: string;
  target_event_id: string;
  event_ids: string[];
  edges: ProjectHistoryPathEdge[];
  minimum_fused_score: number;
  truth_status_code: "inferred";
  source_relation_code: "post_lineage_edge";
  provenance: "post_lineage_edge.fused_score";
}

export interface ProjectHistoryEvent {
  event_id: string;
  source_post_id: string;
  event_title: string;
  event_type_code: string;
  event_type_basis_code: "controlled_source_code";
  occurred_at: string;
  time_basis_code: ProjectHistoryTimeBasis;
  voc_type_code: string | null;
  source_stage_code: string | null;
  source_detail_state_code: string | null;
  project_matches: ProjectHistoryMatch[];
  observed_responsibilities: ProjectHistoryResponsibility[];
  responsibility_transition_code: ResponsibilityTransitionCode | null;
  related_prior_paths: ProjectHistoryPriorPath[];
}

export interface ProjectHistoryProjection {
  contract_version: 1;
  project_key: string;
  normalized_project_key: string;
  project_name: string;
  focus_event_id: string;
  time_basis_code: ProjectHistoryTimeBasis;
  event_count: number;
  distinct_observed_actor_count: number;
  truncated: boolean;
  events: ProjectHistoryEvent[];
}

function normalizeProjectIdentity(value: string): string {
  return value.normalize("NFKC").trim().toLocaleLowerCase("en-US");
}

/** Return one display key per exact normalized project identity. */
export function projectHistoryKeys(
  evidence: ProjectEvidence[] | undefined,
  sourceProjectCode: string | null | undefined,
  sourceProjectName: string | null | undefined,
): string[] {
  const candidates = evidence?.length
    ? evidence.map((project) => project.project_key)
    : [sourceProjectCode?.trim() ? sourceProjectCode : sourceProjectName ?? ""];
  const seen = new Set<string>();
  return candidates.filter((candidate) => {
    const normalized = normalizeProjectIdentity(candidate);
    if (!normalized || seen.has(normalized)) return false;
    seen.add(normalized);
    return true;
  });
}

const MESSAGE_KEYS = [
  "heading",
  "summaryCounts",
  "documentTime",
  "truncated",
  "eventDetail",
  "eventType",
  "eventDate",
  "timeBasisCode",
  "recordedEventTime",
  "sourceCreationTime",
  "sourceStageCode",
  "sourceDetailStateCode",
  "responsibilityEvidence",
  "noResponsibilityEvidence",
  "continuous",
  "handoff",
  "assignmentGap",
  "priorHistory",
  "noPriorHistory",
  "inferredBoundary",
  "projectEvidence",
  "observed",
  "inferred",
  "openSourceRecord",
  "exactValues",
  "exactTableLabel",
  "columnDate",
  "columnEvent",
  "columnType",
  "columnTransition",
  "columnActors",
  "columnPathScore",
  "notApplicable",
  "contractAwarded",
  "specificationChanged",
  "delivered",
  "handoffRecorded",
  "vocReceived",
  "rebidStarted",
  "sourceRecorded",
] as const;

export const PROJECT_HISTORY_MESSAGE_KEYS = MESSAGE_KEYS;
export type ProjectHistoryMessageKey = (typeof MESSAGE_KEYS)[number];

type MessageParams = Record<string, string | number>;

const EN: Record<ProjectHistoryMessageKey, string> = {
  heading: "Project event timeline",
  summaryCounts: "{events} events · {actors} observed actors",
  documentTime: "Dates use the recorded event time when available and the source creation time otherwise.",
  truncated: "This bounded timeline is truncated. The selected event remains included.",
  eventDetail: "Event detail",
  eventType: "Display event type",
  eventDate: "Event date",
  timeBasisCode: "Time source",
  recordedEventTime: "Recorded event time",
  sourceCreationTime: "Source creation time",
  sourceStageCode: "Source stage code",
  sourceDetailStateCode: "Source detail-state code",
  responsibilityEvidence: "Observed responsibility evidence",
  noResponsibilityEvidence: "No responsibility evidence is recorded for this event.",
  continuous: "Responsibility evidence continued",
  handoff: "Responsibility evidence changed",
  assignmentGap: "Responsibility evidence gap",
  priorHistory: "Related prior history",
  noPriorHistory: "No visible prior lineage path is recorded for this event.",
  inferredBoundary: "This is inferred related history, not causality or an authoritative assignment record.",
  projectEvidence: "Project identity evidence",
  observed: "Observed",
  inferred: "Inferred",
  openSourceRecord: "Open source record: {title}",
  exactValues: "Exact values",
  exactTableLabel: "Project history exact values",
  columnDate: "Date",
  columnEvent: "Event",
  columnType: "Type",
  columnTransition: "Responsibility evidence change",
  columnActors: "Observed actors",
  columnPathScore: "Minimum lineage score",
  notApplicable: "Not applicable",
  contractAwarded: "Contract awarded",
  specificationChanged: "Specification changed",
  delivered: "Delivered",
  handoffRecorded: "Handoff recorded",
  vocReceived: "VOC received",
  rebidStarted: "Rebid started",
  sourceRecorded: "Source record",
};

const MESSAGES: Record<Locale, Record<ProjectHistoryMessageKey, string>> = {
  en: EN,
  ko: {
    heading: "프로젝트 이벤트 타임라인",
    summaryCounts: "이벤트 {events}건 · 관찰된 담당자 {actors}명",
    documentTime: "기록된 사건 시각을 우선 사용하고, 없으면 원천 생성 시각을 사용합니다.",
    truncated: "이 제한된 타임라인은 일부만 표시합니다. 선택한 이벤트는 계속 포함됩니다.",
    eventDetail: "이벤트 상세",
    eventType: "표시용 이벤트 유형",
    eventDate: "이벤트 날짜",
    timeBasisCode: "시간 출처",
    recordedEventTime: "기록된 사건 시각",
    sourceCreationTime: "원천 생성 시각",
    sourceStageCode: "원천 단계 코드",
    sourceDetailStateCode: "원천 세부 상태 코드",
    responsibilityEvidence: "관찰된 담당 근거",
    noResponsibilityEvidence: "이 이벤트에는 기록된 담당 근거가 없습니다.",
    continuous: "담당 근거 유지",
    handoff: "담당 근거 변경",
    assignmentGap: "담당 근거 공백",
    priorHistory: "관련 과거 이력",
    noPriorHistory: "이 이벤트로 이어지는 공개 가능한 이전 계보가 없습니다.",
    inferredBoundary: "이는 추론된 관련 이력이며 인과관계나 권위 있는 인사 배정 기록이 아닙니다.",
    projectEvidence: "프로젝트 식별 근거",
    observed: "관찰됨",
    inferred: "추론됨",
    openSourceRecord: "원천 기록 열기: {title}",
    exactValues: "정확한 값",
    exactTableLabel: "프로젝트 이력 정확한 값",
    columnDate: "날짜",
    columnEvent: "이벤트",
    columnType: "유형",
    columnTransition: "담당 근거 변화",
    columnActors: "관찰된 담당자",
    columnPathScore: "최소 계보 점수",
    notApplicable: "해당 없음",
    contractAwarded: "수주 확정",
    specificationChanged: "사양 변경",
    delivered: "납품",
    handoffRecorded: "인수인계 기록",
    vocReceived: "VOC 접수",
    rebidStarted: "재입찰 시작",
    sourceRecorded: "원천 기록",
  },
  zh: {
    heading: "项目事件时间线",
    summaryCounts: "{events} 个事件 · {actors} 名已观察责任人",
    documentTime: "优先使用已记录的事件时间；若无，则使用来源创建时间。",
    truncated: "此有界时间线已截断，但所选事件仍保留。",
    eventDetail: "事件详情",
    eventType: "显示事件类型",
    eventDate: "事件日期",
    timeBasisCode: "时间来源",
    recordedEventTime: "已记录的事件时间",
    sourceCreationTime: "来源创建时间",
    sourceStageCode: "来源阶段代码",
    sourceDetailStateCode: "来源详细状态代码",
    responsibilityEvidence: "已观察的责任证据",
    noResponsibilityEvidence: "此事件没有记录责任证据。",
    continuous: "责任证据持续",
    handoff: "责任证据变化",
    assignmentGap: "责任证据缺口",
    priorHistory: "相关既往历史",
    noPriorHistory: "此事件没有可见的既往谱系路径。",
    inferredBoundary: "这是推断的相关历史，并非因果关系或权威任命记录。",
    projectEvidence: "项目身份依据",
    observed: "已观察",
    inferred: "已推断",
    openSourceRecord: "打开源记录：{title}",
    exactValues: "精确值",
    exactTableLabel: "项目历史精确值",
    columnDate: "日期",
    columnEvent: "事件",
    columnType: "类型",
    columnTransition: "责任证据变化",
    columnActors: "已观察责任人",
    columnPathScore: "最低谱系分数",
    notApplicable: "不适用",
    contractAwarded: "合同授予",
    specificationChanged: "规格变更",
    delivered: "已交付",
    handoffRecorded: "已记录交接",
    vocReceived: "收到客户之声",
    rebidStarted: "重新投标开始",
    sourceRecorded: "源记录",
  },
  ja: {
    heading: "プロジェクトイベントのタイムライン",
    summaryCounts: "イベント {events}件 · 観察された担当者 {actors}名",
    documentTime: "記録されたイベント時刻を優先し、ない場合は原資料の作成時刻を使います。",
    truncated: "この上限付きタイムラインは省略されていますが、選択イベントは保持されます。",
    eventDetail: "イベント詳細",
    eventType: "表示用イベント種別",
    eventDate: "イベント日",
    timeBasisCode: "時刻の出典",
    recordedEventTime: "記録されたイベント時刻",
    sourceCreationTime: "原資料の作成時刻",
    sourceStageCode: "ソース段階コード",
    sourceDetailStateCode: "ソース詳細状態コード",
    responsibilityEvidence: "観察された担当根拠",
    noResponsibilityEvidence: "このイベントには担当根拠が記録されていません。",
    continuous: "担当根拠が継続",
    handoff: "担当根拠が変更",
    assignmentGap: "担当根拠の空白",
    priorHistory: "関連する過去履歴",
    noPriorHistory: "このイベントに至る可視の過去系譜はありません。",
    inferredBoundary: "これは推論された関連履歴であり、因果関係や権威ある配属記録ではありません。",
    projectEvidence: "プロジェクト識別根拠",
    observed: "観察済み",
    inferred: "推論済み",
    openSourceRecord: "原資料を開く: {title}",
    exactValues: "正確な値",
    exactTableLabel: "プロジェクト履歴の正確な値",
    columnDate: "日付",
    columnEvent: "イベント",
    columnType: "種別",
    columnTransition: "担当根拠の変化",
    columnActors: "観察担当者",
    columnPathScore: "最小系譜スコア",
    notApplicable: "該当なし",
    contractAwarded: "受注確定",
    specificationChanged: "仕様変更",
    delivered: "納品",
    handoffRecorded: "引継ぎ記録",
    vocReceived: "VOC受付",
    rebidStarted: "再入札開始",
    sourceRecorded: "原資料",
  },
  vi: {
    heading: "Dòng thời gian sự kiện dự án",
    summaryCounts: "{events} sự kiện · {actors} người phụ trách được quan sát",
    documentTime: "Ưu tiên thời gian sự kiện đã ghi; nếu thiếu thì dùng thời gian tạo nguồn.",
    truncated: "Dòng thời gian có giới hạn này đã bị rút gọn nhưng vẫn giữ sự kiện đang chọn.",
    eventDetail: "Chi tiết sự kiện",
    eventType: "Loại sự kiện hiển thị",
    eventDate: "Ngày sự kiện",
    timeBasisCode: "Nguồn thời gian",
    recordedEventTime: "Thời gian sự kiện đã ghi",
    sourceCreationTime: "Thời gian tạo nguồn",
    sourceStageCode: "Mã giai đoạn nguồn",
    sourceDetailStateCode: "Mã trạng thái chi tiết nguồn",
    responsibilityEvidence: "Bằng chứng trách nhiệm quan sát được",
    noResponsibilityEvidence: "Không có bằng chứng trách nhiệm được ghi cho sự kiện này.",
    continuous: "Bằng chứng trách nhiệm tiếp tục",
    handoff: "Bằng chứng trách nhiệm thay đổi",
    assignmentGap: "Khoảng trống bằng chứng trách nhiệm",
    priorHistory: "Lịch sử trước đó có liên quan",
    noPriorHistory: "Không có đường dẫn lịch sử trước đó khả kiến cho sự kiện này.",
    inferredBoundary: "Đây là lịch sử liên quan được suy luận, không phải quan hệ nhân quả hay hồ sơ phân công có thẩm quyền.",
    projectEvidence: "Bằng chứng nhận dạng dự án",
    observed: "Đã quan sát",
    inferred: "Đã suy luận",
    openSourceRecord: "Mở bản ghi nguồn: {title}",
    exactValues: "Giá trị chính xác",
    exactTableLabel: "Giá trị chính xác của lịch sử dự án",
    columnDate: "Ngày",
    columnEvent: "Sự kiện",
    columnType: "Loại",
    columnTransition: "Thay đổi bằng chứng trách nhiệm",
    columnActors: "Người phụ trách được quan sát",
    columnPathScore: "Điểm dòng dõi tối thiểu",
    notApplicable: "Không áp dụng",
    contractAwarded: "Đã trao hợp đồng",
    specificationChanged: "Đã thay đổi đặc tả",
    delivered: "Đã bàn giao sản phẩm",
    handoffRecorded: "Đã ghi nhận bàn giao",
    vocReceived: "Đã nhận ý kiến khách hàng",
    rebidStarted: "Đã bắt đầu đấu thầu lại",
    sourceRecorded: "Bản ghi nguồn",
  },
};

export function projectHistoryText(
  locale: Locale,
  key: ProjectHistoryMessageKey,
  params: MessageParams = {},
): string {
  let value = MESSAGES[locale][key];
  for (const [name, replacement] of Object.entries(params)) {
    value = value.replaceAll(`{${name}}`, String(replacement));
  }
  return value;
}

export function projectHistoryEventTypeLabel(locale: Locale, code: string): string {
  const keyByCode: Record<string, ProjectHistoryMessageKey> = {
    contract_awarded: "contractAwarded",
    specification_changed: "specificationChanged",
    delivered: "delivered",
    handoff_recorded: "handoffRecorded",
    voc_received: "vocReceived",
    rebid_started: "rebidStarted",
    source_recorded: "sourceRecorded",
  };
  const key = keyByCode[code];
  return key ? projectHistoryText(locale, key) : code;
}

export function projectHistoryTransitionLabel(
  locale: Locale,
  code: ResponsibilityTransitionCode | null,
): string {
  if (code === "continuous") return projectHistoryText(locale, "continuous");
  if (code === "handoff") return projectHistoryText(locale, "handoff");
  if (code === "assignment_gap") return projectHistoryText(locale, "assignmentGap");
  return projectHistoryText(locale, "notApplicable");
}
